crates/bitcoind_rpc/src/bip158.rs

```diff
@@ -6,262 +6,218 @@
+                    cp = cp.range(..=height).next().expect("we found a base");
+
+                    header
                 }
-            })
-        })()
-        .transpose()
-    }
-}
+            };
 
-impl<C: RpcApi> FilterIter<'_, C> {
-    /// Returns the point of agreement between `self` and the given `cp`.
-    fn find_base_with(&mut self, mut cp: CheckPoint) -> Result<BlockId, Error> {
-        loop {
-            let height = cp.height();
-            let fetched_hash = match self.blocks.get(&height) {
-                Some(hash) => *hash,
-                None if height == 0 => cp.hash(),
-                _ => self.client.get_block_hash(height as _)?,
+            let mut next_hash = match header.next_block_hash {
+                Some(hash) => hash,
+                None => return Ok(None),
             };
-            if cp.hash() == fetched_hash {
-                // ensure this block also exists in self
-                self.blocks.insert(height, cp.hash());
-                return Ok(cp.block_id());
+
+            let mut next_header = self.client.get_block_header_info(&next_hash)?;
+
+            // In case of a reorg, rewind by fetching headers of previous hashes until we find
+            // one with enough confirmations.
+            let mut reorg_ct: i32 = 0;
+            while next_header.confirmations < 0 {
+                let prev_hash = next_header
+                    .previous_block_hash
+                    .ok_or(Error::ReorgDepthExceeded)?;
+                let prev_header = self.client.get_block_header_info(&prev_hash)?;
+                next_header = prev_header;
+                reorg_ct += 1;
             }
-            // remember conflicts
-            self.blocks.insert(height, fetched_hash);
-            cp = cp.prev().expect("must break before genesis");
-        }
-    }
 
-    /// Returns a chain update from the newly scanned blocks.
-    ///
-    /// Returns `None` if this [`FilterIter`] was not constructed using a [`CheckPoint`], or
-    /// if no blocks have been fetched for example by using [`get_tip`](Self::get_tip).
-    pub fn chain_update(&mut self) -> Option<CheckPoint> {
-        if self.cp.is_none() || self.blocks.is_empty() {
-            return None;
-        }
+            next_hash = next_header.hash;
+            let next_height: u32 = next_header.height.try_into()?;
 
-        // note: to connect with the local chain we must guarantee that `self.blocks.first()`
-        // is also the point of agreement with `self.cp`.
-        Some(
-            CheckPoint::from_block_ids(self.blocks.iter().map(BlockId::from))
-                .expect("blocks must be in order"),
-        )
+            // Purge any no longer valid checkpoints.
+            if reorg_ct.is_positive() {
+                cp = cp
+                    .range(..=next_height)
+                    .next()
+                    .ok_or(Error::ReorgDepthExceeded)?;
```

Comment:
We'll need to keep track of the heights we have removed to ensure the checkpoint update connects.